nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/IdentifyMimeType.java

```diff
@@ -206,34 +236,50 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
 
         final ComponentLog logger = getLogger();
         final AtomicReference<String> mimeTypeRef = new AtomicReference<>(null);
+        final AtomicReference<String> extensionRef = new AtomicReference<>(null);
         final String filename = flowFile.getAttribute(CoreAttributes.FILENAME.key());
 
         session.read(flowFile, new InputStreamCallback() {
             @Override
             public void process(final InputStream stream) throws IOException {
                 try (final InputStream in = new BufferedInputStream(stream);
                      final TikaInputStream tikaStream = TikaInputStream.get(in)) {
-                    Metadata metadata = new Metadata();
 
-                    if (filename != null && context.getProperty(USE_FILENAME_IN_DETECTION).asBoolean()) {
-                        metadata.add(TikaCoreProperties.RESOURCE_NAME_KEY, filename);
+                    if (customDetector != null) {
+                        detectMimeType(customDetector, customMimeTypes, tikaStream);
+                    }
+
+                    String mimeType = mimeTypeRef.get();
+
+                    if (defaultDetector != null && (mimeType == null || mimeType.equals(MediaType.OCTET_STREAM.toString()) || mimeType.equals(MediaType.TEXT_PLAIN.toString()))) {
+                        detectMimeType(defaultDetector, defaultMimeTypes, tikaStream);
                     }
-                    // Get mime type
-                    MediaType mediatype = detector.detect(tikaStream, metadata);
-                    mimeTypeRef.set(mediatype.toString());
                 }
             }
+
+            private void detectMimeType(Detector detector, MimeTypes mimeTypes, TikaInputStream tikaStream) throws IOException {
+                Metadata metadata = new Metadata();
+                if (filename != null && context.getProperty(USE_FILENAME_IN_DETECTION).asBoolean()) {
+                    metadata.add(TikaCoreProperties.RESOURCE_NAME_KEY, filename);
+                }
+
+                MediaType mediatype = detector.detect(tikaStream, metadata);
+                String extension = "";
+
+                try {
+                    MimeType mimetype = mimeTypes.forName(mediatype.toString());
+                    extension = mimetype.getExtension();
+                } catch (MimeTypeException ex) {
+                    logger.warn("MIME type extension lookup failed: {}", new Object[]{ex});
```

Comment:
The `Object[]` wrapper is deprecated and should not be used.
```suggestion
                    logger.warn("MIME type extension lookup failed", e);
```

Reply:
done